src/qibo/gates/gates.py

```diff
@@ -2,11 +2,102 @@
 from typing import List
 
 import numpy as np
+import sympy as sp
 
 from qibo.config import PRECISION_TOL, raise_error
 from qibo.gates.abstract import Gate, ParametrizedGate
 
 
+class Parameter:
+    """Object which allows for variational gate parameters. Several trainable parameter
+    and possibly features are linked through a lambda function which returns the
+    final gate parameter
+
+    Args:
+        func (function): lambda function describing the gate parameter
+        trainable (list or np.ndarray): array with initial trainable parameters theta
+        feature (list or np.ndarray): array containing possible input features x
+    """
+
+    def __init__(self, func, trainable, feature=None):
+        self._trainable = trainable
+        self._feature = feature
+        self.nparams = len(trainable)
+
+        if isinstance(feature, list):
+            self.nfeat = len(feature)
+        else:
+            self.nfeat = 0
+        self.lambdaf = func
+        self.derivatives = self.calculate_derivatives()
+
+    def _apply_func(self, function, fixed_params=None):
+        """Applies lambda function and returns final gate parameter"""
+        params = []
+        if self._feature is not None:
+            if isinstance(self._feature, list):
+                params.extend(self._feature)
+            else:
+                params.append(self._feature)
+        if fixed_params:
+            params.extend(fixed_params)
+        else:
+            params.extend(self._trainable)
+        return float(function(*params))
+
+    def update_parameters(self, trainable=None, feature=None):
+        """Update gate trainable parameter and feature values"""
+        if not isinstance(trainable, (list, np.ndarray)):
+            raise_error(
+                ValueError, "Trainable parameters must be given as list or numpy array"
+            )
+
+        if not isinstance(trainable, (list, np.ndarray)):
+            raise_error(ValueError, "Features must be given as list or numpy array")
+
+        if trainable is not None:
+            self._trainable = trainable
+        if feature and self._feature:
+            self._feature = feature
+
+    def get_gate_parameters(self):
+        """Update values with trainable parameter and calculate current gate parameter"""
+        return self._apply_func(self.lambdaf)
+
+    def get_indices(self, start_index):
+        """Return list of respective indices of trainable parameters within
+        the optimizer's trainable parameter list"""
+        return [start_index + i for i in range(self.nparams)]
+
+    def get_fixed_part(self, trainable_idx):
+        """Retrieve parameter constant unaffected by a specific trainable parameter"""
+        params = self._trainable.copy()
+        params[trainable_idx] = 0.0
+        return self._apply_func(self.lambdaf, fixed_params=params)
+
+    def calculate_derivatives(self):
```

Comment:
Did you to adopt this approach in order not to re-calculate every time the derivatives? 

I think it can be useful to have a simpler function in which the derivative w.r.t whatever in the lambda can be calculated individually. One can be interested, e.g., in derivating the expected value w.r.t only one of the features.

Reply:
When you initialise the Parameter, it automatically calculates all possible internal derivatives. Then, based on the parameter you want to the know the partial derivative of, you just call Parameter.get_partial_derivative() (previously called get_scaling_factor). This function therefore already exists :)

Reply:
This is exactly what I was thinking about. When we do QML stuff, we update the parameters many times! Think to the scenario in which one use a CMA optimizer in a QML algo: if we do like you suggest, the derivatives are calculated every time a parameter is updated and never used! I don't really know heavy is the differentiation, but I still believe a safest solution is to calculate when is necessary.

Reply:
So I think it depends a bit if the actual Parameter lambda changes a lot. If it's the same function, say theta1 * x + theta2, then the derivatives are calculated once when initialising the Parameter (you get derivatives [theta1, x, 1] for all three possible parameters) and when you update either a feature or parameter, those functions are already in memory and you simply plugin the new values into the derivative formulas. So as long as the lambda function has the same "shape", we don't need to recalculate the derivatives

Reply:
Oh, I see. These are callable functions, so we simply have the analytical form of the derivative. That souns good, thanks for the explanation.